### DISCUSSION TITLE
Disable converting device code from LLVM IR to SPIRV

### DISCUSSION THREAD

**Discussion by harishch4** [intel/llvm#8368] [Q&A]
While reading about SYCL Device compiler [here](https://intel.github.io/llvm-docs/design/CompilerAndRuntimeDesign.html), it is mentioned that 

> Middle-end -
> 
>     Optionally: LLVM IR → SPIR-V translator.

Howerver, by default SPIR-V translation seems to be enabled. How can I disable it?

**Comment by bader** [comment]
> While reading about SYCL Device compiler [here](https://intel.github.io/llvm-docs/design/CompilerAndRuntimeDesign.html), it is mentioned that
> 
> > Middle-end -
> > ```
> > Optionally: LLVM IR → SPIR-V translator.
> > ```
> 
> Howerver, by default SPIR-V translation seems to be enabled. How can I disable it?

In this context, SPIR-V is not enabled by default. Middle-end output (LLVM IR) is stored in fat object file by default.

What exactly are you trying to accomplish by disabling SPIR-V translator?

**Reply by harishch4** [reply]
> In this context, SPIR-V is not enabled by default. Middle-end output (LLVM IR) is stored in fat object file by default.

Okay, Thanks.

> What exactly are you trying to accomplish by disabling SPIR-V translator?

My understanding is that the fat binary contains SPIR-V code and the device OpenCL compiler coverts the SPIR-V code to device binary. What happens if the device OpenCL implementation doesn't support ch_khr_spir extension?

**Reply by bader** [reply]
OpenCL implementations support device code either in source code format (e.g. OpenCL C, C++ for OpenCL) or binary (SPIR-1.2, SPIR-V or native binary).
If implementation supports `cl_khr_spir`, it can consume SPIR-1.2. https://man.opencl.org/cl_khr_spir.html. NOTE: it's not SPIR-V. Technically, SPIR-1.2 is based on LLVM 3.2, which is very old version of LLVM IR. Full specification - https://registry.khronos.org/SPIR/specs/spir_spec-1.2.pdf.

> What happens if the device OpenCL implementation doesn't support ch_khr_spir extension?

If OpenCL implementation doesn't support SPIR-V format, there is no guarantee that SYCL applications built with DPC++ compiler will work correctly.

**Reply by harishch4** [reply]
> OpenCL implementations support device code either in source code format (e.g. OpenCL C, C++ for OpenCL) or binary (SPIR-1.2, SPIR-V or native binary).

Then how does the LLVM IR in fat binary get compiled to device? 

> NOTE: it's not SPIR-V. Technically, SPIR-1.2 is based on LLVM 3.2, which is very old version of LLVM IR. 

Thanks for the clarification. I'm not aware of this. 

>OpenCL implementations support device code either in source code format (e.g. OpenCL C, C++ for OpenCL) or binary (SPIR-1.2, SPIR-V or native binary).

What is the right way to check if OpenCL implementation supports SPIR-V?

**Reply by bader** [reply]
> > OpenCL implementations support device code either in source code format (e.g. OpenCL C, C++ for OpenCL) or binary (SPIR-1.2, SPIR-V or native binary).
> 
> Then how does the LLVM IR in fat binary get compiled to device?

LLVM IR stored in fat objects, but when fat objects are linked to fat binary device code format is transformed to SPIR-V. SPIR-V is transferred to the OpenCL.

> 
> > NOTE: it's not SPIR-V. Technically, SPIR-1.2 is based on LLVM 3.2, which is very old version of LLVM IR.
> 
> Thanks for the clarification. I'm not aware of this.
> 
> > OpenCL implementations support device code either in source code format (e.g. OpenCL C, C++ for OpenCL) or binary (SPIR-1.2, SPIR-V or native binary).
> 
> What is the right way to check if OpenCL implementation supports SPIR-V?

Check if `cl_khr_il_program` is supported for OpenCL versions before 2.1. 2.1 later versions are required to support SPIR-V.

NOTE: Some DPC++ feature requires non-standard SPIR-V extensions.

**Reply by harishch4** [reply]
Got it, Thanks!!
